_posts/2018-10-23-artsy-auth.md

```diff
@@ -0,0 +1,279 @@
+worry about the implementation - assume that you have access to a function called `encrypt(message, secret)`. So if
+we want to encrypt our message, we can say `encrypt(MSG, ABC)` and we'll get back a string of meaningless
+characters, let's call it `ENC`. The key here is that there also exists a function `decrypt(encrypted, secret)`,
+such that as long as the secret is the same will convert your encrypted string back into its original state.
+`decrypt(ENC, ABC) === MSG`
+
+This is basically how security works on much of the internet, with various twists that optimize for different
+things depending on what exactly is being secured and from whom. If you want to send a message out into the world
+but you only want a specific person to be able to read it, you can encrypt it with your secret and then share the
+secret with only those people that you want to allow to read your message.
+
+This is what's going on with our `Authorization` token. Gravity generates a JWT which encodes some facts about the
+user or application under discussion. It then encrypts that JWT using a secret key, and attaches the encrypted
+token to the HTTP request using the `Authorization` header. You've got that same secret key in your application.
+This allows your app to decrypt the value of the `Authorization` token and get at the JWT that it contains.
+
+So what's _that_ look like?
+
+### JWT Details
+
+A JWT (or JSON Web Token) is a small JSON object that stores information about some subject. It has a small set of
+pre-defined fields, some of which we use, as well as support for arbitrary custom fields.
+
+When Gravity generates a JWT for a user it
+[specifies four fields 🔒](https://github.com/artsy/gravity/blob/master/app/models/util/user_trust.rb#L29-L34):
+
+1. `sub` is part of the JWT spec and is short for `subject` - this encodes the subject of the token, in other words
+   the user on whose behalf it's been generated.
+2. `salt_hash` is a unique identifier for the given user with the given password within the gravity context.
+3. `roles` is a string representing a comma-delimited list of the roles that the user has within the Artsy
+   authentication structure.
+4. `partner_ids` holds references to any Partner objects to which the user has access.
+
+When Gravity generates a JWT for an application, on the other hand, it
+[only specifies roles 🔒](https://github.com/artsy/gravity/blob/master/app/models/util/application_trust.rb#L14).
+There is no subject - the fact that it was encoded using a secret relevant to our application is sufficient grounds
+to know that it's about our application. Does that make sense? We don't need to specify a subject because if it was
+generated for any other application we'd never be able to decrypt it. So we just need to know what roles this
+application has been granted.
+
+So when we decrypt the JWT in our application we get either a token that represents a trusted user or a token that
```

Comment:
sorry - it's a style nit, but you start two sentences with "so" in a row ( I do this all the time, and thus am really cognizant of it)